### PR TITLE
Mark unsafe functions unsafe, as Ferris intended

### PR DESCRIPTION
**Pull request by domenukk**

### CODE REVIEW

File: libafl_qemu/src/modules/calls.rs [hunk 1]
```diff
@@ -543,19 +543,19 @@ static mut CALLSTACKS: Option<ThreadLocal<UnsafeCell<Vec<GuestAddr>>>> = None;
...
         unsafe { (*addr_of_mut!(CALLSTACKS)) = Some(ThreadLocal::new()) };
         Self {}
     }
 
     pub fn reset(&mut self) {
+        // # Safety
+        // This accesses the global [`CALLSTACKS`] variable.
+        // While it is racey, it might be fine if multiple clear the vecs concurrently.
+        // TODO: This should probably be rewritten in a safer way.
         unsafe {
```
Comment (domenukk) [inline libafl_qemu/src/modules/calls.rs:559]: This is actually shady and could probably be rewritten in a safer way? Technically this function needs to be marked unsafe, but it propagates up a lot.. cc @rmalmain

Comment (rmalmain) [reply libafl_qemu/src/modules/calls.rs:559]: yes, [this is definitiely easy to break](https://godbolt.org/z/zM8Gexn1v)
we should just return a full vector (not a ref) at this point; it's cloned later on anyway in every caller of the `backtrace` function.

**Comment by domenukk** [comment]
@rmalmain any objections to the rest of the PR?

File: libafl_qemu/src/emu/hooks.rs [hunk 2]
```diff
@@ -1221,8 +1245,10 @@ where
         self.hooks.syscalls_function(hook)
     }
 
+    /// # Safety
+    /// Calls through to the, potentially unsafe, `syscalls_closure`
     #[allow(clippy::type_complexity)]
-    pub fn syscalls_closure(
+    pub unsafe fn syscalls_closure(
```
Comment (domenukk) [inline libafl_qemu/src/emu/hooks.rs:1251]: @rmalmain would you say the closure are actually all safe to call?

**Comment by rmalmain** [comment]
looks goot to me.
not sure the atomic is really necessary on the `BREAK_ON_TMOUT`, but it's not happening in a hotpath so it shouldn't have any performance impact.

**Review by rmalmain — APPROVED** [review]

File: libafl_qemu/src/emu/hooks.rs (hunk 2 above)
Comment (rmalmain) [reply libafl_qemu/src/emu/hooks.rs:1251]: i'd say yes (at least as of today).
rust implements Box<dyn Fn> with a fat pointer, so the transmute *should* never unsafe.
in practice, i don't think the language guarantees anything about this anywhere, it just happens to work with the current implementation of closures.

Comment (domenukk) [reply libafl_qemu/src/emu/hooks.rs:1251]: so should I remove the `unsafe` thing here?

Comment (rmalmain) [reply libafl_qemu/src/emu/hooks.rs:1251]: hum, do we consider it's safe to assume the compiler will keep this implementation the same internally in the future? no sure what to think there tbh

Comment (domenukk) [reply libafl_qemu/src/emu/hooks.rs:1251]: Is there any way to (debug_)assert this?

Comment (rmalmain) [reply libafl_qemu/src/emu/hooks.rs:1251]: not really imho, it's a semantic thing.
transmute compiles only if both types have the same size, maybe we can consider this enough?

Comment (domenukk) [reply libafl_qemu/src/emu/hooks.rs:1251]: So I'll remove the unsafe again and we YOLO it?

**Pull request merged**